File: selenium-agent-mcp-starter-python/src/pages/base_page.py
from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.remote.webelement import WebElement
from src.config.settings import get_env


class BasePage:
    def __init__(self, driver: WebDriver):
        self.driver = driver
        self.explicit_wait = int(get_env("EXPLICIT_WAIT", "10"))

    def navigate_to(self, url: str):
        self.driver.get(url)

    def find_element(self, locator: tuple):
        return self.driver.find_element(*locator)

    def find_elements(self, locator: tuple):
        return self.driver.find_elements(*locator)

    def wait_for_element(self, locator: tuple, timeout: int = None):
        timeout = timeout or self.explicit_wait
        return WebDriverWait(self.driver, timeout).until(
            EC.presence_of_element_located(locator)
        )

    def wait_for_visible_element(self, locator: tuple, timeout: int = None):
        timeout = timeout or self.explicit_wait
        return WebDriverWait(self.driver, timeout).until(
            EC.visibility_of_element_located(locator)
        )

    def click(self, locator: tuple):
        element = WebDriverWait(self.driver, self.explicit_wait).until(
            EC.element_to_be_clickable(locator)
        )
        element.click()

    def type(self, locator: tuple, text: str):
        element = self.wait_for_visible_element(locator)
        element.click()
        element.send_keys(Keys.CONTROL, "a")
        element.clear()
        element.send_keys(text)

    def get_text(self, locator: tuple) -> str:
        element = self.wait_for_element(locator)
        return element.text

    def get_attribute(self, locator: tuple, name: str) -> str:
        element = self.wait_for_element(locator)
        return element.get_attribute(name)

    def is_visible(self, locator: tuple) -> bool:
        return self.wait_for_visible_element(locator).is_displayed()

    def current_url(self) -> str:
        return self.driver.current_url

    def title(self) -> str:
        return self.driver.title

    def click_element(self, element: WebElement):
        self.driver.execute_script("arguments[0].scrollIntoView({block: 'center'});", element)
        element.click()

    def js_click(self, locator: tuple):
        element = self.wait_for_visible_element(locator)
        self.driver.execute_script("arguments[0].scrollIntoView({block: 'center'});", element)
        self.driver.execute_script("arguments[0].click();", element)

    def wait_for_url_contains(self, value: str, timeout: int = None):
        timeout = timeout or self.explicit_wait
        WebDriverWait(self.driver, timeout).until(EC.url_contains(value))

    def take_screenshot(self, filepath: str):
        self.driver.save_screenshot(filepath)

File: selenium-agent-mcp-starter/pages/base.page.ts
import { By, WebDriver, WebElement, until } from 'selenium-webdriver';

export class BasePage {
  constructor(protected driver: WebDriver) {}

  async navigate(url: string): Promise<void> {
    await this.driver.get(url);
  }

  async findElement(locator: By): Promise<WebElement> {
    return this.driver.findElement(locator);
  }

  async waitForElement(locator: By, timeout = 5000): Promise<WebElement> {
    return this.driver.wait(until.elementLocated(locator), timeout);
  }

  async waitForVisible(locator: By, timeout = 5000): Promise<WebElement> {
    const element = await this.waitForElement(locator, timeout);
    await this.driver.wait(until.elementIsVisible(element), timeout);
    return element;
  }

  async click(locator: By): Promise<void> {
    const element = await this.waitForVisible(locator);
    await element.click();
  }

  async jsClick(locator: By): Promise<void> {
    const element = await this.waitForVisible(locator);
    await this.driver.executeScript('arguments[0].scrollIntoView({block: "center"});', element);
    await this.driver.executeScript('arguments[0].click();', element);
  }

  async type(locator: By, text: string): Promise<void> {
    const element = await this.waitForVisible(locator);
    await element.clear();
    await element.sendKeys(text);
  }

  async getText(locator: By): Promise<string> {
    const element = await this.waitForVisible(locator);
    return element.getText();
  }

  async getCurrentUrl(): Promise<string> {
    return this.driver.getCurrentUrl();
  }

  async getTitle(): Promise<string> {
    return this.driver.getTitle();
  }

  async isVisible(locator: By): Promise<boolean> {
    try {
      const element = await this.waitForVisible(locator);
      return element.isDisplayed();
    } catch {
      return false;
    }
  }

  async getAttribute(locator: By, attribute: string): Promise<string> {
    const element = await this.waitForVisible(locator);
    return (await element.getAttribute(attribute)) ?? '';
  }

  async getElements(locator: By): Promise<WebElement[]> {
    return this.driver.findElements(locator);
  }

  async waitForUrlContains(path: string, timeout = 5000): Promise<void> {
    await this.driver.wait(until.urlContains(path), timeout);
  }

  async takeScreenshot(filePath: string): Promise<string> {
    const image = await this.driver.takeScreenshot();
    // In a real implementation, write the base64 image to a file
    return image;
  }
}
